Blockly enhancements 1 -

stefan-hoehn opened the pull request:
This is the first step into adding many more openhab blockly blocks. 

- add a structured menu with sub groups: audio, voice, timers, variables, notifications
- add code documentation, tooltips and help urls
- refactoring

Details can be found in the following thread:
https://community.openhab.org/t/extending-blockly-with-new-openhab-commands/127169

hmerk commented:
Hi Stefan,
sorry, but I am no git expert, so can't really tell.
I think you need to bring your main branch "up2date" and then rebase your enhancement branch...

stefan-hoehn commented:
@asmigala Thanks for helping me - very much appreciated.

I struggle with deletion of the "three lines"... (see at the end)

This is what I do

- I have cloned my repo again into a new directory to make sure I am not losing anything of what I have developed so far

`git clone https://github.com/stefan-hoehn/openhab-webui/
git remote add upstream https://github.com/openhab/openhab-webui.git
`


- Then I do as you say

`git checkout blockly_enhancements_1
git fetch origin
`

- To see where I am at I show the last commits with "tig" (I added the commit hashes)

`
cffa127 2021-10-16 13:02 +0200 Stefan Höhn        o [blockly_enhancements_1] {origin/blockly_enhancements_1} [blockly] add new blocklies: audio, v

789b72a 2021-04-10 07:39 +0200 Stefan Höhn        o [main] {origin/HEAD} {origin/main} allow width/height for floorplan marker as advanced params


001160b  2021-04-07 14:51 +0200 Stefan Höhn        o allow width/height for floorplan marker as advanced params(fix typo)
9c4079c 2021-04-07 11:46 +0200 Stefan Höhn        o allow width/height for floorplan marker as advanced params

ebd4b04 2021-04-02 15:52 +0200 Wouter Born        o Upgrade Karaf to 4.3.1 (#984)

`

I'd be happy to loose the last four commits as I have a copy of the newest and the other 3 can be lost

Now I do 

`git rebase -i origin/main
`
and the editor shows:

pick cffa127c [blockly] add new blocklies: audio, voice, timer

\# Rebase 789b72ae..cffa127c onto 789b72ae (1 command)
\#

I have no experience with rebase but reading the manual, doesn't have this to be changed to include the last one ebd4b04 like so?

#not necessarily this one but would be okay 
drop cffa127 

\# but these
drop 789b72a
drop 001160b
drop 9c4079c
pick ebd4b04

or (again, I really don't know what I am doing ;-)

`Rebase ebd4b04..cffa127c onto ebd4b04
`

Thanks again for you help, 
Stefan

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_ephemeris.js:
```diff
@@ -0,0 +1,128 @@
+  Blockly.Blocks['oh_ephemeris_basic'] = {
+    init: function () {
+      this.appendDummyInput()
+        .appendField(new Blockly.FieldDropdown([['isWeekend', 'isWeekend'], ['getBankHolidayName', 'getBankHolidayName'], ['getNextBankHoliday', 'getNextBankHoliday'], ['isBankHoliday', 'isBankHoliday']]), 'type')
```
These should have proper natural language labels:
- it is the weekend
- it is a bank holiday

I don't understand what this block is for, depending on what's selected in the dropdown, it retrieves a String value a boolean. It's clearly not the right way of designing blocks.
Probably best to add individual blocks like below (which I don't see in the toolbox)

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_ephemeris.js:
```diff
@@ -0,0 +1,128 @@
+    return [code, Blockly.JavaScript.ORDER_NONE]
+  }
+
+  Blockly.Blocks['oh_Ephemeris_getBankHolidayName'] = {
```
This doesn't appear in the toolbox.
Also please use consistent casing (lowercase snake_case).

stefan-hoehn replied:
good catch - these would have been the individual ones that you had asked for above only that haven't been put into the editor file (but without the offset support). As mentioned above if we did it this way, we would end up in 8 different blocks.

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_timers.js:
```diff
@@ -0,0 +1,326 @@
+  * Simple Timer creation without further control of retriggering, canceling or the like
+  * Blockly part
+  */
+  Blockly.Blocks['oh_simpleTimer'] = {
```
oh_timer_simple

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_timers.js:
```diff
@@ -0,0 +1,326 @@
+  Blockly.Blocks['oh_timer_item'] = {
+    init: function () {
+      this.appendDummyInput()
+        .appendField('Timer Name')
```
"timer" is enough

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_timers.js:
```diff
@@ -0,0 +1,326 @@
+    init: function () {
+      this.appendStatementInput('timer')
+        .setCheck(null)
+        .appendField('After')
```
after

Make sure all labels in Blockly are lower case and read like a natural languagge sentence (i.e. no technical function names etc., see the other existing blocks for inspiration).

stefan-hoehn replied:
Sure, found some more.

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_timers.js:
```diff
@@ -0,0 +1,326 @@
+  * Simple Timer creation without further control of retriggering, canceling or the like
+  * Code part
+  */
+  Blockly.JavaScript['oh_simpleTimer'] = function (block) {
```
oh_timer_simple

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_timers.js:
```diff
@@ -0,0 +1,326 @@
+        .setCheck(null)
+        .appendField('After')
+        .appendField(new Blockly.FieldNumber(10), 'delay')
+        .appendField(new Blockly.FieldDropdown([['seconds', 'seconds'], ['minutes', 'minutes'], ['hours', 'hours']]), 'time_unit')
```
timeUnit

(use pascalCase for fields and snake_case for block IDs).

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_timers.js:
```diff
@@ -0,0 +1,326 @@
+    addScriptExecution()
+    addZonedDateTime()
+    let numberDelay = block.getFieldValue('delay')
+    let dropdownTimeUnit = block.getFieldValue('time_unit')
```
timeUnit

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_timers.js:
```diff
@@ -0,0 +1,326 @@
+  * Creates a named timer that starts after the defined delay provided my the number and the unit of time
+  * Blockly part
+  */
+  Blockly.Blocks['oh_namedTimer'] = {
```
oh_timer

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_timers.js:
```diff
@@ -0,0 +1,326 @@
+      this.appendValueInput('delay')
+        .setCheck(null)
+        .appendField('After')
+        .appendField(new Blockly.FieldDropdown([['seconds', 'plusSeconds'], ['minutes', 'plusMinutes'], ['hours', 'plusHours'], ['days', 'plusDays'], ['weeks', 'plusWeeks'], ['months', 'plusMonths']]), 'delayUnits')
```
I prefer this one to the timeUnit field of the oh_timer_simple above so perhaps change the implementation of the simple timer to have exactly the same parameter?

stefan-hoehn replied:
You mean adding the days, weeks, months to the simpleTimer as they are missing there?

![image](https://user-images.githubusercontent.com/5937600/138288918-37872ce7-a886-4739-884f-33e54b376e72.png)

like so?

![image](https://user-images.githubusercontent.com/5937600/138289074-dd22e2ef-fa94-4563-8d39-cfa087a1ba8a.png)

ghys replied:
yes have the same options/field name, also avoid the numerical field and use a value input with a Number type constraint instead for the delay (the "10" in the screenshots above), that way you can use a variable or the result of a calculation instead of a predefined value.

ghys commented on bundles/org.openhab.ui/web/src/components/widgets/plan/oh-plan-marker.vue:
```diff
@@ -11,7 +11,8 @@
         {{ tooltip }}
       </div>
-      <oh-icon v-else-if="config.icon" :icon="config.icon" :color="config.iconColor" :width="config.iconWidth || config.iconSize || 40" :height="config.iconHeight || config.iconSize || 40" :state="config.iconUseState ? state : undefined" />
+      <oh-icon v-else-if="config.icon && config.icon.indexOf('oh:') === 0" :icon="config.icon.split(':')[1]" :width="config.iconWidth || config.iconSize || 40" :height="config.iconHeight || config.iconSize || 40" :state="config.iconUseState ? state : undefined" />
```
Revert changes to this file.

stefan-hoehn replied:
See comment above on how to convert this correctly...

ghys commented on bundles/org.openhab.ui/web/src/pages/settings/rules/script/blockly-editor.vue:
```diff
@@ -328,156 +266,388 @@
+            </value>
+          </block>
+        </category>
+        <category name="Voice/Multimedia">
```
Voice & Multimedia

(to be consistent)

ghys commented on bundles/org.openhab.ui/web/src/pages/settings/rules/script/blockly-editor.vue:
```diff
@@ -328,156 +266,388 @@
+            </value>
+          </block>
+        </category>
+        <category name="Logging/Output">
```
Logging & Output

ghys commented on bundles/org.openhab.ui/web/src/pages/settings/rules/script/blockly-editor.vue:
```diff
@@ -328,156 +266,388 @@
-      stroke inherit
-.blocklyDropDownDiv
-  z-index 9000
+.blockly-editor {
```
Please revert back to the Stylus syntax.

ghys replied:
I mean the style changes without the {} and : and ; where perfectly fine as they were. Stylus is a simplified CSS syntax.

ghys commented on bundles/org.openhab.ui/web/src/pages/settings/rules/script/blockly-editor.vue:
```diff
@@ -328,156 +266,388 @@
 import Vue from 'vue'
 
 import defineOHBlocks from '@/assets/definitions/blockly/ohblocks'
+import defineOHBlocksTimers from '@/assets/definitions/blockly/ohblocks_timers'
```
Maybe these imports should be moved to an index.js file in `@/assets/definitions/blockly/index.js`, with a function to define all blocks in one go.

stefan-hoehn replied:
I wouldn't mind but as you are the expert on this I would recommend you rather do after we are done with all the other changes, so I don't break anything?

ghys commented on bundles/org.openhab.ui/web/src/pages/settings/rules/script/blockly-editor.vue:
```diff
@@ -328,156 +266,388 @@
-    this.workspace.removeChangeListener(this.onChange)
+  created () {
+    this.getAltData()
+    this.getAddons()
```
What is this for?

stefan-hoehn replied:
This is used to retrieve the information from the backend to have the voices, scripts, rules and addons available to provide them as lookups in the dropdowns.

ghys commented on bundles/org.openhab.ui/web/src/pages/settings/rules/script/blockly-editor.vue:
```diff
@@ -328,156 +266,388 @@
+      voices: [],
+      scripts: [],
+      rules: [],
+      addons: [],
```
What is this for?

stefan-hoehn replied:
This are used for the providing the options in the dropdowns.
Note that scripts is not used yet because it is one of the block-groups that I haven't worked on but is coming soon (see the thread post of me and Rich about bringing this up), so i would leave that in for now.
Regarding the addons I am not really sure why "bigbased" had put that in - I currently cannot see a usage for that so we might be able to remove that but it would mean I have to do a full regression test on all blocks to make sure it has no side effect taking it out.

ghys commented on bundles/org.openhab.ui/web/src/pages/settings/rules/script/blockly-editor.vue:
```diff
@@ -328,156 +266,388 @@
+    },
+    loadPage () {
+      defineOHBlocks(this.$f7)
+      defineOHBlocksTimers(this.$f7)
```
See comment above.

ghys commented on bundles/org.openhab.ui/web/src/pages/settings/rules/script/blockly-editor.vue:
```diff
@@ -328,156 +266,388 @@
+
+      this.startBlockly()
+    },
+    getAddons () {
```
What is this for?

stefan-hoehn replied:
As mentioned above this seems to be the only unused method but as I took over the code I am not really sure why it was put it. If I take it out we need to do a complete regression test.

ghys replied:
I would remove this, I don't think it's used anywhere.

ghys commented on bundles/org.openhab.ui/web/src/pages/settings/rules/script/blockly-editor.vue:
```diff
@@ -328,156 +266,388 @@
     getCode () {
       return Blockly.JavaScript.workspaceToCode(this.workspace)
-    },
-    onChange (event) {
```
Where has this gone?

stefan-hoehn replied:
I just readded it.

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_timers.js:
```diff
@@ -0,0 +1,326 @@
+    addGlobalVar(timerName)
+
+    let code = `if (this.${timerName} === null) {\n`
+    code += `\tthis.${timerName} = scriptExecution.createTimer(zonedDateTime.now().${delayunits}(${delay}), function(){\n`
```
I believe Blockly indents code with 4 spaces, not tabs?

ghys replied:
Also I don't like having global variables defined for every timer, would it be possible to put them all under an object, i.e. `this.timers.${timerName}`?
(Naturally this object would have to be initialized and checked for nullness in other blocks).

stefan-hoehn replied:
Regarding the indenting I am not sure about tabs or 4 spaces. I can do that if you want (tabs though are nicer in the code I think). However, keep in mind, the indenting is not perfect, in particular in multi-lines and code within code. It is better than nothing but don't expect perfectly generated readable code (at least I have not found any hints during research how to do that in a better way).

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_timers.js:
```diff
@@ -0,0 +1,326 @@
+  * Creates a named timer that starts after the defined delay provided my the number and the unit of time
+  * Code part
+  */
+  Blockly.JavaScript['oh_namedTimer'] = function (block) {
```
oh_timer

stefan-hoehn commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks.js:
```diff
@@ -1,133 +1,112 @@
       this.appendDummyInput()
         .appendField('item')
         .appendField(new FieldItemModelPicker('MyItem', null, { f7 }), 'itemName')
-      this.setColour(0)
```
Fair statement, it had taken it over from the original source and not really taken care of it. I changed it to (0) which is the base color of the OH blocks. The only thing I would mention though is that we now have about 25+ openhab blocks that all have the same color where differentiation of subgroups with a somehow different color (e.g for timers) might make sense.

stefan-hoehn commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks.js:
```diff
@@ -1,133 +1,112 @@
-        .setCheck('String')
+      this.appendDummyInput()
+        .appendField('thing')
+        .appendField(new FieldItemThingPicker('MyThing', null, { f7 }), 'itemName')
```
Done, I also refactored the name in the code and as a result all code-expressions of all blocks in that class by using string interpolation.

stefan-hoehn commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks.js:
```diff
@@ -1,133 +1,112 @@
-    let code = 'itemRegistry.getItem(' + itemName + ').getState()'
+  Blockly.JavaScript['oh_thing'] = function (block) {
+    const itemName = block.getFieldValue('itemName')
+    let code = '\'' + itemName + '\''
```
Yep, great minds think alike :) see comment above (hadn't read this one here yet :-) )

stefan-hoehn commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks.js:
```diff
@@ -1,133 +1,112 @@
+      this.appendDummyInput()
+        .appendField('thing')
+        .appendField(new FieldItemThingPicker('MyThing', null, { f7 }), 'itemName')
+      this.setColour(248)
```
Done for ALL blocks now.

stefan-hoehn commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks_audio.js:
```diff
@@ -0,0 +1,276 @@
+      this.setPreviousStatement(true, null)
+      this.setNextStatement(true, null)
+      this.setColour(230)
+      this.setTooltip('Plays a sound file from the sounds folder to the given sink at a given volume. \n Note: rather set volume first via thing volume channel, then play sound without volume parameter as it may be more reliable..')
```
Very thorough review :)

stefan-hoehn commented on bundles/org.openhab.ui/web/src/pages/settings/rules/script/blockly-editor.vue:
```diff
@@ -328,156 +266,388 @@
+            </value>
+          </block>
+        </category>
+        <category name="Timers and Delays">
```
Again very good catch - I only saw this lately and was already too tired ;-)

stefan-hoehn commented:
@ghys Can you check my two comments at https://community.openhab.org/t/extending-blockly-with-new-openhab-commands/127169/44

stefan-hoehn commented on bundles/org.openhab.ui/web/src/assets/definitions/blockly/ohblocks.js:
```diff
@@ -1,133 +1,112 @@
   }
 
-  Blockly.Blocks['oh_log'] = {
+  Blockly.Blocks['oh_getitem'] = {
```
ok, already on it, I maybe I have that implemented soon enough and it it to the PR... ;-)

stefan-hoehn commented:
This should be hopefully the finally push for the enhancement. Believe me, I tried several times to rebase and squash the commits but failed badly (I pay someone a 🍺 if s/he is explaining me how to do this right 😢 ).

So @ghys, can you please see if this all can now be merged? Thanks

bigbasec commented:
Not sure that this is even still all valid and working.   I worked on this a looong time ago, but never got any feedback or anything, so I sort of dropped it.   It's not something I ever intended to use anyway.   I can probably still help here a bit, but not sure what needs to be done still.

stefan-hoehn commented:
hey @bigbasec - good to see you finally. You provided a really good foundation you can be proud of on which I have based my work on. Without that I wouldn't have gotten that far so quickly. I refactored, fixed a lot, made it work, extended it and added quite some code and tested it. Now I hope we can release it very soon and everyone can use it. Nothing to do here at the moment anymore but you really deserve the appreciation to have it all started. If you like to see what happened in the last weeks, you can look [here](https://community.openhab.org/t/extending-blockly-with-new-openhab-commands/127169).

ghys commented:
Just so you know - I'll make sure I will credit you both equally in the final commit as you've both made sizeable contributions.

stefan-hoehn commented:
@ghys Just to be sure (and not meaning to nag you) - Do you need anything from me to merge the PR? I am done with my work so far.

ghys reviewed (approved):
Okay so I finished this myself, these last commits include a lot of adjustments & refactoring so make sure you pull the latest changes from main and start from there if you want to continue adding more blocks.

Also updated to the latest **stable** Blockly (Q2 2021, 6.20210701.0) - see https://github.com/google/blockly/releases for the changelog.
Noticed the upcoming release has a lot of changes, including the ability to define the toolbox using JSON instead of XML, which looks interesting to simplify the dynamic creation with the (upcoming) user-defined Blockly libraries.
